homeassistant/components/automation/__init__.py

```diff
@@ -215,6 +232,109 @@ async def reload_service_handler(service_call):
     return True
 
 
+class AutomationTrace:
+    """Container for automation trace."""
+
+    def __init__(self, unique_id, config, trigger, context, action_trace):
+        """Container for automation trace."""
+        self._action_trace = action_trace
+        self._condition_trace = None
+        self._config = config
+        self._context = context
+        self._error = None
+        self._state = "running"
+        self._timestamp_finish = None
+        self._timestamp_start = dt_util.utcnow()
+        self._trigger = trigger
+        self._unique_id = unique_id
+        self._variables = None
+
+    def set_error(self, ex):
+        """Set error."""
+        self._error = ex
+
+    def set_variables(self, variables):
+        """Set variables."""
+        self._variables = variables
+
+    def set_condition_trace(self, condition_trace):
+        """Set condition trace."""
+        self._condition_trace = condition_trace
+
+    def finished(self):
+        """Set finish time."""
+        self._timestamp_finish = dt_util.utcnow()
+        self._state = "stopped"
```

Comment:
Would it help if we froze the instance when this happens (so raise in mutating methods if state is stopped)

Reply:
This is called from the `finally` clause of the `trace_automation` context manager though, not sure where and how to do the test?